server/app/views/admin/programs/ProgramBlockView.java

```diff
@@ -1,98 +1,844 @@
+
+    FormTag createRepeatedBlockForm =
+        form(csrfTag)
+            .withId(CREATE_REPEATED_BLOCK_FORM_ID)
+            .withMethod(HttpVerbs.POST)
+            .withAction(blockCreateAction)
+            .with(
+                FieldWithLabel.number()
+                    .setFieldName(ENUMERATOR_ID_FORM_FIELD)
+                    .setScreenReaderText(ENUMERATOR_ID_FORM_FIELD)
+                    .setValue(OptionalLong.of(blockId))
+                    .getNumberTag());
+
+    String blockDeleteAction =
+        controllers.admin.routes.AdminProgramBlocksController.destroy(programId, blockId).url();
+    FormTag deleteBlockForm =
+        form(csrfTag)
+            .withId(DELETE_BLOCK_FORM_ID)
+            .withMethod(HttpVerbs.POST)
+            .withAction(blockDeleteAction);
+
+    return div(createBlockForm, createRepeatedBlockForm, deleteBlockForm).withClasses("hidden");
+  }
+
+  private DivTag blockOrderPanel(Request request, ProgramDefinition program, long focusedBlockId) {
+    DivTag ret = div().withClasses("shadow-lg", "pt-6", "w-2/12", "border-r", "border-gray-200");
+    ret.with(
+        renderBlockList(
+            request,
+            program,
+            program.getNonRepeatedBlockDefinitions(),
+            focusedBlockId,
+            /* level= */ 0));
+
+    if (viewAllowsEditingProgram()) {
+      ret.with(
+          ViewUtils.makeSvgTextButton("Add screen", Icons.ADD)
+              .withClasses(AdminStyles.SECONDARY_BUTTON_STYLES, "m-4")
+              .withType("submit")
+              .withId("add-block-button")
+              .withForm(CREATE_BLOCK_FORM_ID));
+    }
+    return ret;
+  }
+
+  private DivTag renderBlockList(
+      Request request,
+      ProgramDefinition programDefinition,
+      ImmutableList<BlockDefinition> blockDefinitions,
+      long focusedBlockId,
+      int level) {
+    DivTag container = div().withClass("pl-" + level * 2);
+    for (BlockDefinition blockDefinition : blockDefinitions) {
+
+      // TODO: Not i18n safe.
+      int numQuestions = blockDefinition.getQuestionCount();
+      String questionCountText = String.format("Question count: %d", numQuestions);
+      String blockName = blockDefinition.name();
+
+      String selectedClasses = blockDefinition.id() == focusedBlockId ? "bg-gray-100" : "";
+      DivTag blockTag =
+          div()
+              .withClasses(
+                  "flex",
+                  "flex-row",
+                  "gap-2",
+                  "py-2",
+                  "px-4",
+                  "border",
+                  "border-white",
+                  StyleUtils.hover("border-gray-300"),
+                  selectedClasses);
+      String editUrl =
+          controllers.admin.routes.AdminProgramBlocksController.edit(
+                  programDefinition.id(), blockDefinition.id())
+              .url();
+      String switchBlockLink =
+          viewAllowsEditingProgram() ? editUrl : NOT_YET_IMPLEMENTED_ERROR_MESSAGE;
```

Comment:
**issue (blocking):** Throw an exception here rather than rendering an invalid link.  Though I think you can just assume everything is for editing in this PR given the constructor requires it, rather than doing this handling through the method.   It's fine to keep the `viewAllowsEditingProgram` checks elsewhere

Reply:
See other comment. I think this is good enough as my other Pr is ready.